pkg/softtainter/softtainter.go

```diff
@@ -0,0 +1,507 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	prometheusAuthTokenSecretKey        = "prometheusAuthToken"
+	overutilizedSoftTaintKey            = "nodeutilization.descheduler.kubernetes.io/overutilized"
+	overutilizedSoftTaintValue          = "true"
+	appropriatelySoftTaintKey           = "nodeutilization.descheduler.kubernetes.io/appropriate"
+	appropriatelyutilizedSoftTaintValue = "true"
+)
+
+var (
+	log = logf.Log.WithName("controller_nodeClassifier")
+)
+
+type softTainterArgs struct {
+	useDeviationThresholds bool
+	thresholds             api.ResourceThresholds
+	targetThresholds       api.ResourceThresholds
+	applySoftTaints        bool
+	mode                   desv1.Mode
+}
+
+type softTainter struct {
+	args                              *softTainterArgs
+	resourceNames                     []corev1.ResourceName
+	previousPrometheusClientTransport *http.Transport
+	currentPrometheusAuthToken        string
+	promClient                        promapi.Client
+	promQuery                         string
+	promURL                           string
+	client                            client.Client
+	resyncPeriod                      time.Duration
+	policyConfigFile                  string
+	nodeUtilizationFactory            func(promapi.Client, string) NodeUtilization
+}
+
+type NodeUtilization interface {
+	NodeUsageFromPrometheusMetrics(ctx context.Context) (map[string]map[corev1.ResourceName]*resource.Quantity, error)
+}
+
+type ActualNodeUtilization struct {
+	promClient promapi.Client
+	promQuery  string
+}
+
+func (anu *ActualNodeUtilization) NodeUsageFromPrometheusMetrics(ctx context.Context) (map[string]map[corev1.ResourceName]*resource.Quantity, error) {
+	return nodeutilization.NodeUsageFromPrometheusMetrics(ctx, anu.promClient, anu.promQuery)
+}
+
+func (st *softTainter) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+
+	des := desv1.KubeDescheduler{}
+
+	descheduler.SetupPlugins()
+	policy, err := descheduler.LoadPolicyConfig(st.policyConfigFile, nil, pluginregistry.PluginRegistry)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	err = st.client.Get(ctx, client.ObjectKey{
+		Namespace: operatorclient.OperatorNamespace,
+		Name:      operatorclient.OperatorConfigName,
+	}, &des)
+	if err != nil {
+		logger.Error(err, "failed reading descheduler operator CR")
+		return reconcile.Result{}, err
+	}
+
+	if des.Spec.DeschedulingIntervalSeconds == nil {
+		return reconcile.Result{}, fmt.Errorf("descheduler should have an interval set")
+	}
+	st.resyncPeriod = time.Duration(*des.Spec.DeschedulingIntervalSeconds) * time.Second
+
+	var lnargs *nodeutilization.LowNodeUtilizationArgs
+
+	for _, p := range policy.Profiles {
+		if p.Name == string(desv1.LongLifecycle) || p.Name == string(desv1.RelieveAndMigrate) {
```

Comment:
We can limit running the soft tainter only for `RelieveAndMigrate` profile. Everyone who wants to run the new controller will be asked to set the new profile. Including anyone using the `LongLifecycle` profile. So we limit the support and encourage users to adopt the new profile.

Reply:
done